### PR TITLE
petsc: cleanup test code to support 'spack test run' and add a 'cu…

### PR DESCRIPTION
**Pull request by balay**
and add a 'cuda' test

cc: @jczhang07

### CODE REVIEW

**Comment by adamjstewart** [comment]
@balay why did you remove the requested reviews for @jedbrown and @BarrySmith? They're listed as maintainers for this package, so presumably they would be the best people to review this PR.

**Comment by balay** [comment]
> @balay why did you remove the requested reviews for @jedbrown and @BarrySmith? They're listed as maintainers for this package, so presumably they would be the best people to review this PR.

As I'm the primary maintainer of this package - I was looking for review from spack experts wrt 'test' functionality

Also they are listed more as a FYI - so they are aware of the changes here. If the process now expects them to respond to most petsc related MRs - I'll have to check if they still want to be listed as maintainers.

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/petsc/package.py:519]: Usually we run `make check` after building and before installing. Does this test require the library to be installed first?

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/petsc/package.py:556]: Is this the same for all MPI implementations?

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:519]: yes - for a normal usage - PETSc is considered installed only after 'make install' - so 'make check' is run after that.

And this change did work - when I tested with `spack install --test root petsc`

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:556]: I was hoping there was a spack universal mpiexec equivalent (from the mpi provider that spack picked up) that all mpi related packages could use. Is there a better impl of this - that I should look at?

I checked superlu-dist - that had:
```
           mpiexe_f = which('srun', 'mpirun', 'mpiexec')
            if mpiexe_f:
                self.run_test(mpiexe_f.command, test_args, work_dir='.',
                              purpose='superlu-dist smoke test')
```
I wasn't sure if `which` picks up stuff from PATH. And if an incompatible mpiexec can be in PATH.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:556]: [or if `srun` gets picked up - when `mpiexec` is more appropriate]

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/petsc/package.py:556]: I don't use MPI much. I know we have things like `spec['mpi'].mpicc` for the compiler, but I don't think we have anything for the runner. I'm guessing we probably should. @alalazo @becker33?

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:556]: @adamjstewart perhaps we can merge this in - and update when there is better common support running mpi tests.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
